### PR TITLE
Merge overloads via default parameter values

### PR DESCRIPTION
**Pull request by drewnoakes**
Single commit extracted from PR #4615 on request of @sharwell for separate review.

Changes proposed in this pull request:
 - Reduce the number of method overloads by using default parameter values

What did I do to test the code and ensure quality:
 - Manual review of changes (largely automated)
 - Unit tests
 - Manual testing

Has been tested on:
 - GIT 2.16
 - Windows 10

### CODE REVIEW

File: Plugins/ReleaseNotesGenerator/HtmlFragment.cs [hunk 1]
```diff
@@ -157,27 +157,16 @@ internal static string To8DigitString(int x)
...
         }
 
-        /// <summary>
-        /// Clears clipboard and copy a HTML fragment to the clipboard. This generates the header.
-        /// </summary>
-        /// <param name="htmlFragment">A html fragment.</param>
-        /// <example>
-        ///    HtmlFragment.CopyToClipboard("<b>Hello!</b>");
-        /// </example>
-        public static void CopyToClipboard(string htmlFragment)
```
Comment (drewnoakes) [inline Plugins/ReleaseNotesGenerator/HtmlFragment.cs]: I opted to remove this documentation as it didn't add anything over its sibling.

**Comment by sharwell** [comment]
The issues I'm typically looking for in this type of change fall into one of two categories:

1. Cases where the default parameter is not needed, such as
    * Cases where one of the overloads never should have existed (bad designs, such as making `CancellationToken` optional for an asynchronous method)
    * Cases where the overload with fewer parameters is used infrequently and it's not burdensome to just require the argument
2. Cases where the default parameter allows calls that never should have occurred before

I don't have time right now to do a complete review of the change, but I know an instance of the second case exists in this (or did when it was submitted in the other PR), so I'll point it out there so you know what I'm talking about. 😄

File: GitUI/GitUICommands.cs [hunk 2]
```diff
@@ -512,7 +512,7 @@ public bool DoActionOnRepo(Func<bool> action)
 
         #region Checkout
 
-        public bool StartCheckoutBranch(IWin32Window owner, string branch, bool remote, string[] containRevisons)
+        public bool StartCheckoutBranch(IWin32Window owner, string branch = "", bool remote = false, string[] containRevisons = null)
```
Comment (sharwell) [inline GitUI/GitUICommands.cs]: :memo: It was previously impossible to call `StartCheckoutBranch` and specify `branch` without specifying `remote`. This increases the number of calling combinations allowed for the method and requires each of the combinations be carefully considered.

File: GitUI/GitUICommands.cs [hunk 3]
```diff
@@ -523,29 +523,19 @@ public bool StartCheckoutBranch(IWin32Window owner, string branch, bool remote,
...
+            return StartCheckoutBranch(null, branch, remote);
         }
 
         public bool StartCheckoutBranch()
         {
-            return StartCheckoutBranch(null, "", false, null);
+            return StartCheckoutBranch(null);
         }
 
         public bool StartCheckoutRemoteBranch(IWin32Window owner, string branch)
```
Comment (sharwell) [inline GitUI/GitUICommands.cs]: 📝 This appears to be another overload of `StartCheckoutBranch` with a different name.

File: GitUI/GitUICommands.cs [hunk 4]
```diff
@@ -1929,7 +1841,8 @@ internal void StartPullRequestsDialog(IRepositoryHostPlugin gitHoster)
         public void StartCreatePullRequest(IWin32Window owner)
         {
             List<IRepositoryHostPlugin> relevantHosts =
-                (from gh in RepoHosts.GitHosters where gh.GitModuleIsRelevantToMe(Module) select gh).ToList();
+                RepoHosts.GitHosters.Where(gh => gh.GitModuleIsRelevantToMe(Module)).ToList();
```
Comment (sharwell) [inline GitUI/GitUICommands.cs]: :memo: Unrelated change

File: GitUI/ProcessOutputTimer.cs [hunk 5]
```diff
@@ -19,21 +19,17 @@ public ProcessOutputTimer(DoOutputCallback doOutput)
             _timer.Tick += _timer_Elapsed;
         }
 
-        public void Start(int interval)
+        public void Start(int interval = 600)
```
Comment (sharwell) [inline GitUI/ProcessOutputTimer.cs]: 💭 I prefer using `TimeSpan` over `int`, which currently requires `TimeSpan?` to use with a default argument. The limitation may be lifted as part of the work on dotnet/csharplang#187.

File: GitUI/GitUICommands.cs (hunk 2 above)
Comment (drewnoakes) [reply GitUI/GitUICommands.cs]: True, though the overload `StartCheckoutBranch(string branch, bool remote)` is never called anywhere anyway 😆

If this is merged I'll make another PR that goes through and removes unused overloads. Or I can add that to this PR if preferred.

File: GitUI/GitUICommands.cs (hunk 3 above)
Comment (drewnoakes) [reply GitUI/GitUICommands.cs]: Yep, it could be inlined into the caller. There are several examples like this, even with overloads of the same name. Would be interesting to see what was left after such inlines.

File: GitUI/GitUICommands.cs (hunk 4 above)
Comment (drewnoakes) [reply GitUI/GitUICommands.cs]: Oops. Can revert but it seems harmless.

File: GitUI/ProcessOutputTimer.cs (hunk 5 above)
Comment (drewnoakes) [reply GitUI/ProcessOutputTimer.cs]: Agree. In #4518 I changed the `AsyncLoader.Delay` from `int` to `TimeSpan`. At the very least the name or docs should indicate the units. Early .NET framework code was bad at this in places (case in point, the WinForms `Timer` class used herein).

**Comment by drewnoakes** [comment]
Agree 100% that there's more opportunity here.

For example, this commit only covers overloads where the trailing arguments differ. Others exist (eg. `InvokeEvent`) that could be rearranged to enable optional args.

There are also some overloads that cannot be merged because one implements an interface method, and there's a chance that those interfaces are worth a revisit.

And many of the overloads appear to be unused anyway so could just be deleted.

`GitUICommands` has ~300 members over ~2400 LOC. Many of those members can be consolidated.

The content of this commit is just the mechanical, and therefore safe, translation which comprises a first step towards reducing the large number of methods a little bit.

I have some ideas to try to bring these numbers down further in the not-too-distant future. I get the sense they'd be better suited to dedicated PRs at that time.

File: NetSpell.SpellChecker/Spelling.cs [hunk 6]
```diff
@@ -619,7 +619,7 @@ public int EditDistance(string source, string target, bool positionPriority)
         /// </remarks>
         public int EditDistance(string source, string target)
         {
-            return EditDistance(source, target, true);
+            return EditDistance(source, target, positionPriority: true);
```
Comment (gerhardol) [inline NetSpell.SpellChecker/Spelling.cs]: Was this a preparation for some other change?

File: GitUI/GitUICommands.cs (hunk 2 above)
Comment (gerhardol) [reply GitUI/GitUICommands.cs]: > Or I can add that to this PR if preferred.

No!

File: NetSpell.SpellChecker/Spelling.cs (hunk 6 above)
Comment (drewnoakes) [reply NetSpell.SpellChecker/Spelling.cs]: No, it is purely cosmetic and snuck in by mistake. Can remove if you like, though it's harmless enough.

Comment (gerhardol) [reply NetSpell.SpellChecker/Spelling.cs]: OK

Comment (RussKie) [reply NetSpell.SpellChecker/Spelling.cs]: Let's keep unrelated changes separate

File: GitUI/GitUICommands.cs (hunk 2 above)
Comment (RussKie) [reply GitUI/GitUICommands.cs]: 🤣

**Comment by RussKie** [comment]
@sharwell any objections?

**Comment by sharwell** [comment]
I'm happy with being able to raise my concerns so people are watching for issues in the future. I'm neutral on the change as a whole.

**Comment by RussKie** [comment]
Drew, there's a merge conflict.

**Comment by drewnoakes** [comment]
@RussKie fixed now, thanks.
